### PR TITLE
Chapter 10

### PR DESCRIPTION
**Pull request by mirulzaei**

### CODE REVIEW

**Comment by mirulzaei** [comment]
ready

File: app/assets/stylesheets/application.scss [hunk 1]
```diff
@@ -15,5 +15,4 @@
  */
  @import 'bootstrap-sprockets';
  @import 'bootstrap';
- @import 'color';
- 
\ No newline at end of file
+ @import 'color';
```
Comment (saobv-2223) [inline app/assets/stylesheets/application.scss]:
```suggestion
 @import 'color';
 
```

File: app/controllers/sessions_controller.rb [hunk 2]
```diff
@@ -18,11 +18,11 @@ def destroy
   end
 
   def current_user
-    if user_id = session[:user_id]
+    if (user_id = session[:user_id])
```
Comment (saobv-2223) [inline app/controllers/sessions_controller.rb]:
```suggestion
    if user_id = session[:user_id]
```

File: app/controllers/sessions_controller.rb [hunk 3]
```diff
@@ -18,11 +18,11 @@ def destroy
...
       @current_user ||= User.find_by id: user_id
-    elsif user_id = cookies.signed[:user_id]
+    elsif (user_id = cookies.signed[:user_id])
```
Comment (saobv-2223) [inline app/controllers/sessions_controller.rb]: remove ()

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,32 +1,90 @@
...
+  before_action :load_user, only: %i(index new create)
+  before_action :correct_user, only: %i(edit update)
+  before_action :admin_user, only: :destroy 
+  
   def show
     @user = User.find_by id: params[:id]
-    return if @user
-
+  end 
+
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]:
```suggestion
  end 
```

File: app/controllers/users_controller.rb [hunk 5]
```diff
@@ -1,32 +1,90 @@
...
     @user = User.find_by id: params[:id]
-    return if @user
-
+  end 
+ 
+    
     flash[:danger] = "User not found!"
     redirect_to root_path
+  def index
+      @users = User.page(params[:page]).per Settings.per_page
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb:15]:
```suggestion
  
  def index
      @users = User.page(params[:page]).per Settings.per_page
```

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -1,32 +1,90 @@
...
+      flash[:success] = t "user.update"
+      redirect_to @user
+    else
+      flash.now[:danger] = t "user.update_fail"
+      render :edit
+    end
+  end
 
+  def edit
+    @user = User.find_by(id: params[:id])
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: remove )(

File: app/controllers/users_controller.rb [hunk 7]
```diff
@@ -1,32 +1,90 @@
...
+      flash[:danger] = t "user.delete_fail"
+    end
+    redirect_to users_path
+  end
 
   private
   
   def user_params
     params.require:(user).permit :name, :email, :password, :password_confirmation
+    params.require(:user).permit User::USERS_PARAMS
```
Comment (saobv-2223) [inline app/controllers/users_controller.rb]: using line 63 or 64, not both

File: app/helpers/application_helper.rb [hunk 8]
```diff
@@ -2,4 +2,5 @@ module ApplicationHelper
   def full_title page_title
     base_title = "Sample App"
     page_title.blank? ? base_title : "#{page_title} | #{base_title}"
+
```
Comment (saobv-2223) [inline app/helpers/application_helper.rb]:
```suggestion
```

File: app/helpers/sessions_helper.rb [hunk 9]
```diff
@@ -35,8 +35,33 @@ def forget user
   end
 
   def log_out
-    forget current_user
-    session.delete :user_id 
+    forget(current_user)
+    session.delete(:user_id)
```
Comment (saobv-2223) [inline app/helpers/sessions_helper.rb]: remove ()

File: app/views/layouts/edit.html.erb [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+<h1>Update your profile</h1>
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= render "form" %>
+    <div class="gravatar_edit">
+      <%= gravatar_for @user %>
+      <%= link_to Settings.links.gravatar %>
+    </div>
+  </div>
+</div>
```
Comment (saobv-2223) [inline app/views/layouts/edit.html.erb]:
```suggestion
</div>

```

File: app/views/layouts/users/_form.html.erb [hunk 11]
```diff
@@ -9,4 +9,4 @@
   <%= f.label :password_confirmation, "Confirmation" %>
   <%= f.password_field :password_confirmation, class: "form-control" %>
   <%= f.submit yield(:button_text), class: "btn btn-primary" %>
-<% end %>
+<% end %>
```
Comment (saobv-2223) [inline app/views/layouts/users/_form.html.erb]:
```suggestion
<% end %>

```

File: app/views/layouts/users/_user.html.erb [hunk 12]
```diff
@@ -3,4 +3,4 @@
   <%= link_to user.name, user %>
   <% if current_user.admin? && !current_user?(user) %> | <%= link_to t "user.delete_user", user, method: :delete, data: { confirm: t "user.confirm" } %>
   <% end %>
-</li>
+</li>
```
Comment (saobv-2223) [inline app/views/layouts/users/_user.html.erb]:
```suggestion
</li>

```

File: app/views/sessions/new.html.erb [hunk 13]
```diff
@@ -15,4 +15,4 @@
     <% end %>
     <p><%= t "new_user?" %><%= link_to t("sign_up_now!"), signup_path %></p>
   </div>
-</div>
+</div>
```
Comment (saobv-2223) [inline app/views/sessions/new.html.erb]:
```suggestion
</div>

```

File: app/views/shared/_error_messages.html.erb [hunk 14]
```diff
@@ -10,4 +10,4 @@
     <% end %>
     </ul>
   </div>
-<% end %>
+<% end %>
```
Comment (saobv-2223) [inline app/views/shared/_error_messages.html.erb]:
```suggestion
<% end %>

```

File: config/locales/en.yml [hunk 15]
```diff
@@ -33,4 +33,51 @@ en:
...
+      title: "About"
+  home:
+    message: "Ruby on Rails"
+    page: "This is the home page for the "
+    sample: " sample application."
+  help:
+    message: "Ruby on Rails Help page"
+    book: "book"
+    page: "Get help on the Ruby on Rails Tutorial at the "
+    sample: "To get help on this sample app, see the "
```
Comment (saobv-2223) [inline config/locales/en.yml]:
```suggestion
    sample: "To get help on this sample app, see the "
    
```

File: config/settings.yml [hunk 16]
```diff
@@ -18,4 +19,5 @@ user:
     regex: !ruby/regexp '/\A[\w+\-.]+@[a-z\d\-.]+\.[a-z]+\z/i'
   password:
     min_length: 6
-    
\ No newline at end of file
+  
+per_page: 10
```
Comment (saobv-2223) [inline config/settings.yml]:
```suggestion
per_page: 10    

```

File: db/seeds.rb [hunk 17]
```diff
@@ -1,7 +1,15 @@
...
+
+99.times do |n|
+name = Faker::Name.name
+email = "example-#{n+1}@railstutorial.org"
+password = "password"
+User.create!(name: name,
+             email: email,
+             password: password,
+             password_confirmation: password)
+end
```
Comment (saobv-2223) [inline db/seeds.rb]:
```suggestion
end

```

**Comment by mirulzaei** [comment]
ready

**Comment by ngocvt-0484** [comment]
merged

**Pull request merged**